tools/tokenize/src/main/scala/org/allenai/nlpstack/tokenize/Ai2Tokenizer.scala

```diff
@@ -0,0 +1,217 @@
+    '#' -> Punctuation,
+    '@' -> Punctuation,
+    '!' -> Punctuation,
+    '~' -> Punctuation,
+    '`' -> Punctuation,
+    ',' -> Punctuation,
+    '\'' -> Punctuation,
+    '"' -> Punctuation,
+    '—' -> Punctuation, // long dash
+    '–' -> Punctuation, // some other dash
+    '’' -> Punctuation,
+    '…' -> Punctuation,
+    '“' -> Punctuation,
+    '”' -> Punctuation,
+    '/' -> Punctuation,
+    '-' -> Dash, // normal dash
+    '?' -> Punctuation,
+    '<' -> Punctuation,
+    '>' -> Punctuation,
+    '‘' -> Punctuation,
+    '•' -> Punctuation,
+    '&' -> Punctuation,
+    '‰' -> Punctuation,
+    '|' -> Punctuation,
+    '=' -> Punctuation,
+    '~' -> Punctuation,
+    '\\' -> Punctuation,
+    '°' -> Punctuation
+  ).withDefaultValue(Other)
+
+  // The code supports replacing one sequence with another (as long as the replacement sequence is
+  // as long or shorter than the pattern sequence). In this version, the replacement sequence is
+  // always of length 1, but hard-coding that assumption has no bearing on performance, and it
+  // doesn't make the code shorter, so I left it this way.
+  private val specialCases = IndexedSeq(
+    IndexedSeq("'", "s") -> "'s", // He's doing the thing.
+    IndexedSeq("'", "re") -> "'re", // You're doing the thing.
+    IndexedSeq("'", "ll") -> "'ll", // I'll do the thing.
+    IndexedSeq("'", "m") -> "'m", // I'm doing the thing.
+    IndexedSeq("'", "d") -> "'d", // You'd do the same thing.
+    IndexedSeq("'", "t") -> "'t", // I can't do the thing.
+    IndexedSeq("'", "ve") -> "'ve", // I should've done the thing.
+    IndexedSeq("e", ".", "g", ".") -> "e.g.",
+    IndexedSeq("i", ".", "e", ".") -> "i.e.",
+    IndexedSeq("a", ".", "m", ".") -> "a.m.",
+    IndexedSeq("p", ".", "m", ".") -> "p.m.",
+    IndexedSeq("et", ".", "al", ".") -> "et.al.",
+    IndexedSeq("ph", ".", "d", ".") -> "ph.d.",
+    IndexedSeq("mr", ".") -> "mr.",
+    IndexedSeq("ms", ".") -> "ms.",
+    IndexedSeq("m", ".", "sc", ".") -> "m.sc.",
+    IndexedSeq("b", ".", "sc", ".") -> "b.sc.",
+    IndexedSeq("b", ".", "a", ".") -> "b.a.",
+    IndexedSeq("etc", ".") -> "etc.",
+    IndexedSeq("u", ".", "s", ".") -> "u.s."
+  ).map {
+      case (pattern, replacement) =>
+        pattern -> IndexedSeq(replacement)
+    }
+
+  require(specialCases.forall {
+    case (pattern, replacement) =>
+      pattern.map(_.length).sum == replacement.map(_.length).sum
+  })
+
+  // This is a requirement of the algorithm that uses specialCases, but it could be fixed relatively
+  // easily.
+  require(specialCases.forall {
+    case (pattern, replacement) =>
+      pattern.length >= replacement.length
+  })
+
+  // These are for "shouldn't", "couldn't", etc.
+  private val shouldCouldWould = Set(
+    "should",
+    "could",
+    "would",
+    "must",
+    "had",
+    "did",
+    "does",
+    "is",
+    "was",
+    "has",
+    "do",
+    "have",
+    "were",
+    "are",
+    "wo", // won't
+    "ca" // can't
+  ).map(_ + "n")
```

Comment:
I added this, but it doesn't go here. The logic at the end only combines tokens. It does not split them.